pyquil/operator_estimation.py

```diff
@@ -479,3 +499,166 @@ def measure_observables(qc: QuantumComputer, tomo_experiment: TomographyExperime
                 expectation=np.asscalar(obs_mean),
                 stddev=np.asscalar(np.sqrt(obs_var)),
             )
+
+
+def terms_diagonal_in_tpb(pauli_a: PauliTerm, pauli_b: PauliTerm):
+    """
+    Test if ``pauli_a`` and ``pauli_b`` are diagonal in each others' "natural" tensor product basis.
+
+    Given some PauliTerm, the 'natural' tensor product basis (tpb) to
+    diagonalize this term is the one which diagonalizes each Pauli operator in the
+    product term-by-term.
+
+    For example, X(1) * Z(0) would be diagonal in the 'natural' tensor product basis
+    {(|0> +/- |1>)/Sqrt[2]} * {|0>, |1>}, whereas Z(1) * X(0) would be diagonal
+    in the 'natural' tpb {|0>, |1>} * {(|0> +/- |1>)/Sqrt[2]}. The two operators
+    commute but are not diagonal in each others 'natural' tpb (in fact, they are
+    anti-diagonal in each others 'natural' tpb). This function tests whether two
+    operators given as PauliTerms are both diagonal in each others 'natural' tpb.
+
+    Note that for the given example of X(1) * Z(0) and Z(1) * X(0), we can construct
+    the following basis which simultaneously diagonalizes both operators:
+
+      -- 2 * |0>' = |0> (|0> + |1>) + |1> (|0> - |1>)
+      -- 2 * |1>' = |0> (|0> + |1>) - |1> (|0> - |1>)
+      -- 2 * |2>' = |0> (|0> - |1>) + |1> (|0> + |1>)
+      -- 2 * |3>' = |0> (-|0> + |1>) + |1> (|0> + |1>)
+
+    In this basis, X Z looks like diag(1, -1, 1, -1), and Z X looks like diag(1, 1, -1, -1)
+
+    :param pauli_a: PauliTerm to check diagonality of in the natural tpb of `pauli_b`
+    :param pauli_b: PauliTerm to check diagonality of in the natural tpb of `pauli_a`
+    :return: Boolean of diagonality in each others natural tpb
+    """
+    overlapping_active_qubits = set(pauli_a.get_qubits()) & set(pauli_b.get_qubits())
+    for qubit_index in overlapping_active_qubits:
+        if (pauli_a[qubit_index] != 'I' and pauli_b[qubit_index] != 'I'
+                and pauli_a[qubit_index] != pauli_b[qubit_index]):
+            return False
+
+    return True
+
+
+def _get_diagonalizing_basis(list_of_pauli_terms: List[PauliTerm]):
```

Comment:
write this more pythonically. contrast with `_validate_all_diagonal_in_tpb`

Reply:
yeah, reduce and lambda inhibit readability.

consider how to merge/replace the functionality in `_validate_all_diagonal_in_tpb` so we don't have two functions that do basically the same thing

also: try to avoid `PauliTerm.from_list`. It has really weird semantics. We should probably deprecate it

Reply:
Changed `_get_diagonalizing_basis` to use `_validate_all_diagonal_in_tpb`, then convert the latter's output to a PauliTerm (maybe we should have something like `PauliTerm.from_dict` if it gets used often enough?)